File: C++/Solution552.java
class Solution552 {
    final int MOD = 1000000007;

    public int checkRecord(int n) {
        int[][][] dp = new int[n + 1][2][3];
        dp[0][0][0] = 1;
        for(int len = 1; len <= n; len++){
            // last: L
            for(int numA = 0; numA < 2; numA++){
                for(int nL = 1; nL < 3; nL++){
                    dp[len][numA][nL] += dp[len - 1][numA][nL - 1];
                    dp[len][numA][nL] %= MOD;
                }
            }
            // last: A
            for(int nL = 0; nL < 3; nL++){
                dp[len][1][0] += dp[len - 1][0][nL];
                dp[len][1][0] %= MOD;
            }
            // last: P
            for(int numA = 0; numA < 2; numA++){
                for(int nL = 0; nL < 3; nL++){
                    dp[len][numA][0] += dp[len - 1][numA][nL];
                    dp[len][numA][0] %= MOD;
                }
            }
        }
        int ans = 0;
        for(int numA = 0; numA < 2; numA++){
            for(int nL = 0; nL < 3; nL++){
                ans += dp[n][numA][nL];
                ans %= MOD;
            }
        }
        return ans;
    }
}

File: C++/Solution552.cpp
#include<iostream>
#include<vector>
#include<algorithm>
#include<unordered_map>
#include<unordered_set>
#include<queue>
using namespace std;

typedef long long ll;
typedef pair<int, int> pii;
typedef vector<int> vi;
typedef vector<vi> vii;
typedef vector<vii> viii;

class Solution552 {
public:
    const int MOD = 1e9 + 7;

    int checkRecord(int n) {
        viii dp(n + 1, vii(2, vi(3, 0)));  // len, num of A, last num of L
        dp[0][0][0] = 1;
        for(int len = 1; len <= n; len++){
            // L
            for(int numA = 0; numA < 2; numA++){
                for(int lastL = 1; lastL < 3; lastL++){
                    dp[len][numA][lastL] = (dp[len][numA][lastL] + dp[len - 1][numA][lastL - 1]) % MOD;
                }
            }
            // P
            for(int numA = 0; numA < 2; numA++){
                for(int lastL = 0; lastL < 3; lastL++){
                    dp[len][numA][0] = (dp[len][numA][0] + dp[len - 1][numA][lastL]) % MOD;
                }
            }
            // A
            for(int lastL = 0; lastL < 3; lastL++){
                dp[len][1][0] = (dp[len][1][0] + dp[len - 1][0][lastL]) % MOD;
            }
        }
        int ans = 0;
        for(int numA = 0; numA < 2; numA++){
            for(int lastL = 0; lastL < 3; lastL++){
                ans += dp[n][numA][lastL];
                ans %= MOD;
            }
        }
        return ans;
    }
};